### PR TITLE
WIP: Introduce a second version of the Holland 2010 model

### PR DESCRIPTION
**Pull request by ThomasRoosli**
As discussed I would like to introduce a second option of how b is calculated in the Holland 2010 model is calculated.
In this WIP pull request, could we discuss on the function descriptions and the naming?
I suggest to add a new function called `_bs_holland_2010_v2` being another option to `_bs_holland_2008` (also used in Holland 2010 in the currently implemented version of Holland 2010).
To select this function you would have to specify the parameter `model='H10_v2'` in `TropCylcone.from_tracks`.
What do you think about these names?

Also consider that we might add additional options at a later stage: e.g. how to calculate the static windfield (so an alternative function to `_stat_holland_2010`, again with another formula also mentioned in Holland 2010). This would then probably be selectable with `model='H10_v3'`

### CODE REVIEW

**Comment by tovogt** [comment]
Thanks! As an alternative approach, I would suggest to have a keyword argument `model_kwargs` that allows to set model-specific option parameters. That's what I did in the new TCRain class: https://github.com/CLIMADA-project/climada_petals/blob/develop/climada_petals/hazard/tc_rainfield.py#L235

File: climada/hazard/trop_cyclone.py
Comment (tovogt) [inline climada/hazard/trop_cyclone.py:1353]:
```suggestion
            si_track["vmax"]**2 * RHO_AIR_SURFACE * np.exp(1)
```

Comment (tovogt) [reply climada/hazard/trop_cyclone.py:1353]: To avoid these kinds of mistakes, I strongly suggest to merge the functions `_bs_holland_2010_v2` and `_B_holland_1980` with appropriate optonal parameters to specify the rho, vmax, and clipping to use.

File: climada/hazard/trop_cyclone.py
Comment (tovogt) [inline climada/hazard/trop_cyclone.py:68]: I'm not convinced that this is a good value. There is a formula for air pressure in Holland et al. 2010:
image
If you set SST to 28°C, you get `rho=1.167` (for ps=1005) or `rho=1.103` (for ps=950). I don't see where Holland et al. 2010 suggest to use the value 1.2. I think that it would be fine to stick to the existing constant `RHO_AIR` also for surface air.

**Comment by tovogt** [comment]
I just pushed my suggested alternative implementation (introducing a `model_kwargs` parameter). In the future, this structure could also be used to implement the alternative static windfield formula from Holland et al. 2010, or to specify a second wind measurement (`v_n`, `r_n`).

**Comment by tovogt** [comment]
I just took the opportunity to implement and push the alternative static windfield formula. The results with that alternative formula are very different from all the other models, which makes me wonder whether Holland might just have made a mistake, and wanted to specify two mathematically equivalent formulations instead.

**Comment by tovogt** [comment]
Doing some more checks (trying to reproduce Fig. 1 from the original Holland et al. 2010 publication, and looking at the quite weak storm 2020034S13063), we found that the x-exponent needs truncation to prevent some unphysical results in both versions of the static windfield formula. I implemented these truncations in the latest commit.

**Comment by peanutfun** [comment]
@ThomasRoosli @tovogt Please let me know

**Comment by ThomasRoosli** [comment]
I was able to double check the implementation with our TC expert Pamela Probst looking at many different storms. The Holland Model is correctly implemented. @peanutfun  yes your help is greatly appreciated.

**Comment by peanutfun** [comment]
I'll review the code then as soon as possible. Thanks for the feedback!

File: climada/hazard/test/test_trop_cyclone.py
Comment (peanutfun) [inline climada/hazard/test/test_trop_cyclone.py:444]: These values are given with a precision of 1e-6. So why is it necessary to set the absolute tolerance to 1e-5? Doesn't this mean there is a computation error in here somewhere?

**Review by peanutfun — CHANGES_REQUESTED** [review]
Thanks for updating the Holland model. But I'm confused: Where's the second version mentioned in the PR?

The changes seem reasonable but they increased the size of an already huge module. Could we maybe split it into two files? From line 1132, there are only private methods. Could we move them into a separate module?

I have a few comments regarding the docstrings and tests. Apart from that, the implementation looks solid!

File: climada/hazard/trop_cyclone.py
Comment (peanutfun) [reply climada/hazard/trop_cyclone.py:1233]: Same applies for the later `_compute_angular_windspeeds` functions

File: climada/hazard/trop_cyclone.py
Comment (ThomasRoosli) [reply climada/hazard/trop_cyclone.py:232]: The doi were already cited further below in private methods, I copied the citations into this docstring as requested.

File: climada/hazard/trop_cyclone.py
Comment (ThomasRoosli) [reply climada/hazard/trop_cyclone.py:257]: For me it is very unclear from the paper if `vmax_from_cen` and `vmax_in_brackets` "has to be combined". I agree they would often be used in the same combination, but we should give the user the freedom to choose themselves. For me it is very hard to describe the effect of the parameter change, because I imagine it is very dependent on the TC event and on other chosen settings... Maybe we can sit together and think about a clarification of this docstring. Otherwise I would leave it as suggest by @tovogt.

File: climada/hazard/trop_cyclone.py
Comment (ThomasRoosli) [reply climada/hazard/trop_cyclone.py:1233]: I added the description of the returned array for functions of the "angular windspeed" family.

File: climada/hazard/trop_cyclone.py
Comment (ThomasRoosli) [reply climada/hazard/trop_cyclone.py:247]: It is new to make this configurable. The chosen default now maintains consistency with the results of older code. I think that should be the case, as we want to make configurable in this MR not define a new default behaviour.

File: climada/hazard/trop_cyclone.py
Comment (ThomasRoosli) [reply climada/hazard/trop_cyclone.py:1196]: I assume what happens is that for the frist track position this is set to zero. If this is true, I do not know why this is done here as it will be done again anyway in Line ~1051: `windfields[0, :, :] = 0`.
I personally would not remove this line here, just to repeat it in 4 different places (the `compute_funs`).  I would leave it as it is.

File: climada/hazard/test/test_trop_cyclone.py
Comment (ThomasRoosli) [reply climada/hazard/test/test_trop_cyclone.py:153]: In my view, this test documents the current outcome of the function for a few (rather randomly selected) centroids. The goal of this test is to alarm a future maintainer of the code if their changes are untintentionally changing the results.
I took over this logic and added the new configurations to the same setup. If there s another justification of this test, I am happy to incorporate it...

File: climada/hazard/test/test_trop_cyclone.py
Comment (ThomasRoosli) [reply climada/hazard/test/test_trop_cyclone.py:444]: Due to formatting some values were given with more or less precision. I reformatted to provide all numbers with 1e-6 precision and then the atol can be set accordingly.

**Comment by ThomasRoosli** [comment]
> Thanks for updating the Holland model. But I'm confused: Where's the second version mentioned in the PR?
> 
> The changes seem reasonable but they increased the size of an already huge module. Could we maybe split it into two files? From line 1132, there are only private methods. Could we move them into a separate module?
> 
> I have a few comments regarding the docstrings and tests. Apart from that, the implementation looks solid!

Thanks @peanutfun  for the review.
Regarding your larger questions:
We discontinued the structure with a "second version" proposed by me and replaced it with an "alternative approach" using `model_kwargs `proposed by tovogt.
I suggest splitting the module in two files in a seperate MR.
I tried to answer your comments below please have a look and resolve if you are satisfied.

File: climada/hazard/trop_cyclone.py
Comment (peanutfun) [reply climada/hazard/trop_cyclone.py:269]: Alright, let's leave it as is.

File: climada/hazard/trop_cyclone.py
Comment (peanutfun) [inline climada/hazard/trop_cyclone.py:1301]: `cyclostrophic` is unused in this model. Adding it as a parameter is detected as issue by the linter, and rightly so. Ideas (not mutually exclusive):
1. As the model is cyclostrophic, calling `_compute_angular_windspeeds_h10(cyclostrophic=False)` should emit a warning, right?
2. We can remove the parameter, add a `**kwargs` argument, and check if `cyclostrophic` appears in there, then emit an appropriate warning.

@ThomasRoosli

**Comment by peanutfun** [comment]
@ThomasRoosli One open discussion. Once resolved, we can merge. Sorry for taking so long 🙇

File: climada/hazard/trop_cyclone.py
Comment (peanutfun) [reply climada/hazard/trop_cyclone.py:1307]: After personal discussion: We will tackle this in a possible follow-up PR

**Review by peanutfun — APPROVED** [review]
Very nice work, @tovogt @ThomasRoosli!

Thanks for the latest addition. I'll add a changelog message and merge then.

**Pull request merged**
